### PR TITLE
fog-core 2.x, fog-json 1.x

### PR DESCRIPTION
**Pull request by lanej**
References #227

### CODE REVIEW

**Comment by geemus** [comment]
Looks good in theory, seems there are a couple breakages though, unfortunately.

File: lib/fog/aws/models/elb/load_balancer.rb [hunk 1]
```diff
@@ -223,16 +223,24 @@ def save
...
         end
+
+        protected
+
+        def all_associations_and_attributes
+          super.merge(
+            'ListenerDescriptions' => attributes['ListenerDescriptions'],
+            'BackendServerDescriptions' => attributes['BackendServerDescriptions'],
+          )
+        end
```
Comment (plribeiro3000) [inline lib/fog/aws/models/elb/load_balancer.rb]: Why do you need this? Maybe we need to improve this method on `fog-core`?

Comment (lanej) [reply lib/fog/aws/models/elb/load_balancer.rb]: Attributes are not registered in by an `attribute` declaration but are used to collections from the underlying attributes hash.

This seems like a reasonable workaround using normal OO methodology.

Comment (plribeiro3000) [reply lib/fog/aws/models/elb/load_balancer.rb]: Got it. I'm not really pointing this as a problem but i'm more interested in knowing the reasons behind it. If it does inflect an in improvement in `fog-core` i would be more then happy to do it myself. =)

Are those attributes not being used by the DSL for a particular reason? Is there something i can improve in `fog-core` to help improve the usage?

**Comment by jaredbeck** [comment]
Copying the test failure here for convenience:

```
      tests/models/compute/volume_tests.rb
      MissingParameter => The request must contain the parameter device (Fog::Compute::AWS::Error)
        /home/travis/build/fog/fog-aws/lib/fog/aws/requests/compute/attach_volume.rb:77:in `attach_volume'
        /home/travis/build/fog/fog-aws/lib/fog/aws/models/compute/volume.rb:127:in `attach'
        /home/travis/build/fog/fog-aws/lib/fog/aws/models/compute/volume.rb:92:in `server='
        tests/models/compute/volume_tests.rb:11:in `block (3 levels) in <top (required)>'
...
```

Question: Is this test failure the result of breaking changes in the updated dependencies (eg. fog-core) or is it a result of the preceding warning:

```
  AWS | credentials (aws) +++++[fog][WARNING] Unable to fetch credentials: Expected(200) <=> Actual(404 Not Found)
```

I'm not seeing any mention of "volumes" in the fog-core [changelog](https://github.com/fog/fog-core/blob/master/changelog.md) There were actually very few changes between 1.45 and 2.0

```
564ea20 geemus     2018-01-03 09:23:05 -0600  (tag: v2.0.0) v2.0.0
f5521c2 Paul..eiro 2017-11-13 15:55:42 -0200  Merge pull request #221 from NARKOZ/mime-types
94f76f7 Niha..asov 2017-11-13 16:27:30 +0400  Don't check mime/types for load error
a32e960 Niha..asov 2017-11-13 12:49:55 +0400  Add missing runtime dependency declaration to gemspec
65e33d0 Paul..eiro 2017-10-27 11:14:16 -0200  Merge pull request #220 from JustMcAfee/master
e7e0657 JustMcAfee 2017-10-27 08:30:41 -0400  Updating net-ssh parameter names
9fabe1c Wesl..eary 2017-10-26 08:36:45 -0500  Merge pull request #219 from tbrisker/ruby2
5c53ee1 Tome..sker 2017-10-24 11:48:27 +0300  Drop Ruby<2 support
1af65f0 Paul..eiro 2017-10-02 17:54:56 -0300  Merge pull request #218 from fog/fix-reload-for-associations
85cfeae Paul..eiro 2017-09-30 15:12:53 -0300  fix association reload
13979be geemus     2017-08-01 11:46:25 -0500  (tag: v1.45.0) v1.45.0
```

**Comment by geemus** [comment]
I don't think fog-core has a notion of volumes at all, so I don't think it would be that. Other changes to fog-core could be causing this, I suppose, but I'm not sure which one specifically. I suppose the association reload fix might relate, but I'm not at all certain about that (none of the others look particularly related).

**Comment by geemus** [comment]
@jaredbeck no worries. I think it is probably unrelated (and it's just a warning). Most likely there is an actual issue in the parameters passed within that attach volumes call.

**Comment by lanej** [comment]
@geemus @plribeiro3000 

got it working.  needed to change the volume interface slightly, but I think it's for the better.

File: lib/fog/aws/models/elb/load_balancer.rb (hunk 1 above)
Comment (plribeiro3000) [reply lib/fog/aws/models/elb/load_balancer.rb]: Hmmm. So you mean an `attribute` that ideally does not belong to this resource but the API expects? Like a concept issue from the API standpoint?

BTW, this is not a blocker. I'm just trying to figure out the use case to see if/how we can improve `fog-core`. My main goal now is to get it to a better state.

**Comment by plribeiro3000** [comment]
LGTM!

File: lib/fog/aws/models/compute/volume.rb [hunk 2]
```diff
@@ -109,22 +97,15 @@ def force_detach
...
         end
 
-        private
-
-        def attachmentSet=(new_attachment_set)
-          merge_attributes(new_attachment_set.first || {})
-        end
-
-        def attach(new_server)
+        def attach(new_server, new_device)
```
Comment (geemus) [inline lib/fog/aws/models/compute/volume.rb]: Could you elaborate on why the new argument is needed? The rest looks good, but I don't think I'm quite groking this. Thanks!

Comment (lanej) [reply lib/fog/aws/models/compute/volume.rb]: Tests asserted this procedure

* `v = service.volumes.create(device: '/dev/sda1', **)`
* `v.server = service.instances.create(**) # requires :device`

In step 1, a device is provided to a volume instance. In step 2, the model calls `Fog.wait_for { ready? }` (implicit reload) and then attaches the server.

The issue is between step 1 and 2, there is a property of the local model (`device = '/dev/sda1'`) that does not match the remote model (`device = nil`).  

On `reload`: 

* `fog-core 1.x`, the `device` attribute retained it's value.  
* `fog-core 2.x`, `device` is set the correct remote value (`nil`).

IMO a `reload` should put the local model in the same state as the remote model.  To enforce that, I've removed the ability for `volume.server = server` to attach volume and instead require calling `volume.attach(server, device)`.

**Comment by jaredbeck** [comment]
Thanks @lanej et al. for your work on this.

**Pull request merged**